File: apps/python/src/algorithm/recursion.py
"""第5章 再帰アルゴリズム"""


def factorial(n: int) -> int:
    """n の階乗を再帰的に計算

    >>> factorial(5)
    120
    """
    if n <= 0:
        return 1
    return n * factorial(n - 1)


def gcd(x: int, y: int) -> int:
    """ユークリッドの互除法で最大公約数を求める

    >>> gcd(22, 8)
    2
    """
    if y == 0:
        return x
    return gcd(y, x % y)


def recursive_sum(n: int) -> int:
    """1 から n までの和を再帰的に計算

    >>> recursive_sum(5)
    15
    """
    if n <= 0:
        return 0
    return n + recursive_sum(n - 1)


def hanoi(n: int, src: str, dst: str, via: str) -> list[tuple[str, str]]:
    """ハノイの塔: n 枚の円盤を src から dst へ via を経由して移動する手順を返す

    >>> hanoi(1, 'A', 'C', 'B')
    [('A', 'C')]
    """
    if n == 1:
        return [(src, dst)]
    moves = []
    moves.extend(hanoi(n - 1, src, via, dst))
    moves.append((src, dst))
    moves.extend(hanoi(n - 1, via, dst, src))
    return moves


def maze_solve(
    maze: list[list[int]],
    row: int,
    col: int,
    goal_row: int,
    goal_col: int,
    visited: set[tuple[int, int]] | None = None,
) -> bool:
    """迷路をバックトラッキングで解く

    maze[r][c] == 0: 通路, 1: 壁
    Returns True if goal is reachable from (row, col).

    >>> maze = [[1,1,1],[1,0,1],[1,1,1]]
    >>> maze_solve(maze, 1, 1, 1, 1)
    True
    """
    if visited is None:
        visited = set()

    if row == goal_row and col == goal_col:
        return True

    rows = len(maze)
    cols = len(maze[0])
    visited.add((row, col))

    for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        nr, nc = row + dr, col + dc
        if (
            0 <= nr < rows
            and 0 <= nc < cols
            and maze[nr][nc] == 0
            and (nr, nc) not in visited
        ):
            if maze_solve(maze, nr, nc, goal_row, goal_col, visited):
                return True

    return False


class EightQueen:
    """8 王妃問題（全組み合わせ列挙）

    各列に 1 個の王妃を配置する組み合わせを全列挙する。
    行・対角線の重複チェックなし。
    """

    def __init__(self):
        self.result = []
        self.__pos = [0] * 8

    def put(self) -> None:
        """現在の盤面を結果に追加"""
        self.result.append(self.__pos[:])

    def set(self, i: int) -> None:
        """i 列目に王妃を配置"""
        for j in range(8):
            self.__pos[i] = j
            if i == 7:
                self.put()
            else:
                self.set(i + 1)


class EightQueen2:
    """8 王妃問題（行制約あり）

    各行・各列に 1 個の王妃を配置する組み合わせを列挙する。
    対角線の重複チェックなし。
    """

    def __init__(self):
        self.result = []
        self.__pos = [0] * 8
        self.__flag = [False] * 8  # 各行に王妃が配置済みかのフラグ

    def put(self) -> None:
        """現在の盤面を結果に追加"""
        self.result.append(self.__pos[:])

    def set(self, i: int) -> None:
        """i 列目の適切な位置に王妃を配置"""
        for j in range(8):
            if not self.__flag[j]:
                self.__pos[i] = j
                if i == 7:
                    self.put()
                else:
                    self.__flag[j] = True
                    self.set(i + 1)
                    self.__flag[j] = False


class EightQueen3:
    """8 王妃問題（行・対角線制約あり）

    各行・各列・各対角線に 1 個の王妃を配置する完全な 8 王妃問題。
    92 通りの解を求める。
    """

    def __init__(self):
        self.result = []
        self.__pos = [0] * 8
        self.__flag_a = [False] * 8   # 各行のフラグ
        self.__flag_b = [False] * 15  # 右上がり対角線のフラグ
        self.__flag_c = [False] * 15  # 右下がり対角線のフラグ

    def put(self) -> None:
        """現在の盤面を結果に追加"""
        self.result.append(self.__pos[:])

    def set(self, i: int) -> None:
        """i 列目の適切な位置に王妃を配置"""
        for j in range(8):
            if (not self.__flag_a[j]
                    and not self.__flag_b[i + j]
                    and not self.__flag_c[i - j + 7]):
                self.__pos[i] = j
                if i == 7:
                    self.put()
                else:
                    self.__flag_a[j] = True
                    self.__flag_b[i + j] = True
                    self.__flag_c[i - j + 7] = True
                    self.set(i + 1)
                    self.__flag_a[j] = False
                    self.__flag_b[i + j] = False
                    self.__flag_c[i - j + 7] = False

File: apps/node/src/algorithm/recursion.ts
/**
 * 第5章 再帰アルゴリズム
 */

/** n の階乗を再帰的に計算 */
export function factorial(n: number): number {
  if (n <= 0) return 1;
  return n * factorial(n - 1);
}

/** ユークリッドの互除法で最大公約数を求める */
export function gcd(x: number, y: number): number {
  if (y === 0) return x;
  return gcd(y, x % y);
}

/** 1 から n までの和を再帰的に計算 */
export function recursiveSum(n: number): number {
  if (n <= 0) return 0;
  return n + recursiveSum(n - 1);
}

/** ハノイの塔: n 枚の円盤を src から dst へ via を経由して移動する手順を返す */
export function hanoi(
  n: number,
  src: string,
  dst: string,
  via: string,
): [string, string][] {
  if (n === 1) return [[src, dst]];
  return [
    ...hanoi(n - 1, src, via, dst),
    [src, dst],
    ...hanoi(n - 1, via, dst, src),
  ];
}

/** 迷路をバックトラッキングで解く（0: 通路, 1: 壁） */
export function mazeSolve(
  maze: number[][],
  row: number,
  col: number,
  goalRow: number,
  goalCol: number,
  visited: Set<string> = new Set(),
): boolean {
  if (row === goalRow && col === goalCol) return true;

  const rows = maze.length;
  const cols = maze[0].length;
  visited.add(`${row},${col}`);

  for (const [dr, dc] of [[-1, 0], [1, 0], [0, -1], [0, 1]]) {
    const nr = row + dr;
    const nc = col + dc;
    if (
      nr >= 0 && nr < rows &&
      nc >= 0 && nc < cols &&
      maze[nr][nc] === 0 &&
      !visited.has(`${nr},${nc}`)
    ) {
      if (mazeSolve(maze, nr, nc, goalRow, goalCol, visited)) return true;
    }
  }
  return false;
}

// ─── 8 王妃問題 ────────────────────────────────────────────

/** 8 王妃問題（全組み合わせ列挙） */
export class EightQueen {
  result: number[][] = [];
  private pos: number[];
  private n: number;

  constructor(n: number = 8) {
    this.n = n;
    this.pos = new Array(n).fill(0);
  }

  private put(): void {
    this.result.push([...this.pos]);
  }

  set(i: number): void {
    for (let j = 0; j < this.n; j++) {
      this.pos[i] = j;
      if (i === this.n - 1) this.put();
      else this.set(i + 1);
    }
  }
}

/** 8 王妃問題（行制約あり） */
export class EightQueen2 {
  result: number[][] = [];
  private pos: number[];
  private flag: boolean[];
  private n: number;

  constructor(n: number = 8) {
    this.n = n;
    this.pos = new Array(n).fill(0);
    this.flag = new Array(n).fill(false);
  }

  private put(): void {
    this.result.push([...this.pos]);
  }

  set(i: number): void {
    for (let j = 0; j < this.n; j++) {
      if (!this.flag[j]) {
        this.pos[i] = j;
        if (i === this.n - 1) {
          this.put();
        } else {
          this.flag[j] = true;
          this.set(i + 1);
          this.flag[j] = false;
        }
      }
    }
  }
}

/** 8 王妃問題（行・対角線制約あり）— 92 通りの解 */
export class EightQueen3 {
  result: number[][] = [];
  private pos: number[] = new Array(8).fill(0);
  private flagA: boolean[] = new Array(8).fill(false);
  private flagB: boolean[] = new Array(15).fill(false);
  private flagC: boolean[] = new Array(15).fill(false);

  private put(): void {
    this.result.push([...this.pos]);
  }

  set(i: number): void {
    for (let j = 0; j < 8; j++) {
      if (!this.flagA[j] && !this.flagB[i + j] && !this.flagC[i - j + 7]) {
        this.pos[i] = j;
        if (i === 7) {
          this.put();
        } else {
          this.flagA[j] = true;
          this.flagB[i + j] = true;
          this.flagC[i - j + 7] = true;
          this.set(i + 1);
          this.flagA[j] = false;
          this.flagB[i + j] = false;
          this.flagC[i - j + 7] = false;
        }
      }
    }
  }
}
